File: src/nimble/widgets/rect.cpp
#include <nimble/common.h>
#include <nimble/widgets/rect.h>

#include <layout.h>
#include <nanovg.h>

na::RectWidget::RectWidget(Application* app)
	: Widget(app)
{
}

na::RectWidget::~RectWidget()
{
}

void na::RectWidget::Load(LayoutNode &node)
{
	Widget::Load(node);

	int width = node.GetInt("width", false, m_size.x);
	int height = node.GetInt("height", false, m_size.y);
	SetSize(glm::ivec2(width, height));

	SetColor(node.GetColor("color", false, m_color));
}

void na::RectWidget::DoLayout(lay_context* l, lay_id parent)
{
	lay_id id = BeginLayout(parent);
	lay_set_size_xy(l, id, m_size.x, m_size.y);

	Widget::DoLayout(l, id);
}

void na::RectWidget::Draw(NVGcontext* vg)
{
	if (m_color.a > 0.0f) {
		glm::ivec4 rect = GetLayout();

		nvgSave(vg);
		nvgBeginPath(vg);
		nvgRect(vg, rect.x, rect.y, rect.z, rect.w);
		nvgFillColor(vg, nvgRGBAf(m_color.r, m_color.g, m_color.b, m_color.a));
		nvgFill(vg);
		nvgRestore(vg);
	}

	Widget::Draw(vg);
}

void na::RectWidget::SetSize(const glm::ivec2 &size)
{
	if (m_size != size) {
		InvalidateLayout();
	}
	m_size = size;
}

void na::RectWidget::SetColor(const glm::vec4 &color)
{
	if (m_color != color) {
		InvalidateRendering();
	}
	m_color = color;
}

File: Nimble.App/Widgets/Rect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Nimble.App.Serialization;
using Nimble.App.Math;
using System.Diagnostics;

namespace Nimble.App.Widgets
{
	[WidgetTag("vbox", "hbox")]
	public class Rect : Widget
	{
		private Vector2Int m_size;
		private Vector4 m_color;

		public Rect()
		{
			Console.WriteLine("Measuring");
			Measure();
		}

		public void Measure()
		{
			var sw = new Stopwatch();
			sw.Start();
			for (int i = 0; i < 1000; i++) {
				InvalidateLayout();
			}
			sw.Stop();
			Console.WriteLine("1000: {0}", sw.Elapsed.TotalMilliseconds);
		}

		public override void Load(ContentNode node)
		{
			base.Load(node);

			int width = node.GetInt("width", false, m_size.X);
			int height = node.GetInt("height", false, m_size.Y);
			m_size = new Vector2Int(width, height);

			m_color = node.GetColor("color", false, m_color);
		}

		public Vector2Int Size
		{
			get { return m_size; }
			set {
				if (m_size != value) {
					InvalidateLayout();
				}
				m_size = value;
			}
		}

		public Vector4 Color
		{
			get { return m_color; }
			set {
				if (m_color != value) {
					InvalidateRendering();
				}
				m_color = value;
			}
		}
	}
}
